internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,729 @@
+package servers
+
+import (
+	"context"
+	"crypto/x509"
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/servers/store"
+	"github.com/hashicorp/boundary/internal/types/scope"
+	"github.com/hashicorp/go-dbw"
+	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
+	nodee "github.com/hashicorp/nodeenrollment"
+	"github.com/hashicorp/nodeenrollment/types"
+	"google.golang.org/protobuf/proto"
+)
+
+// Ensure we implement the Storage interfaces
+var (
+	_ nodee.Storage              = (*WorkerAuthRepositoryStorage)(nil)
+	_ nodee.TransactionalStorage = (*WorkerAuthRepositoryStorage)(nil)
+)
+
+// WorkerAuthRepositoryStorage is the Worker Auth database repository
+type WorkerAuthRepositoryStorage struct {
+	reader                db.Reader
+	writer                db.Writer
+	kms                   *kms.Kms
+	worker                *store.Worker
+	workerAuthTransaction *WorkerAuthTransaction
+}
+
+type WorkerAuthTransaction struct {
+	transactionWriter *db.Db
+	transaction       *dbw.RW
+}
+
+// NewRepositoryStoragecreates a new Worker Auth WorkerAuthRepositoryStorage that implements the Storage interface
+// If used to authenticate and store a workerAuth record, a worker must be passed
+// If transactional is set to true, all operations conducted in the lifetime of this storage instance are executed
+// within the context of a single transaction; Flush must be called to either commit or rollback this transaction
+func NewRepositoryStorage(ctx context.Context, r db.Reader, w db.Writer, kms *kms.Kms, wrk *store.Worker, transactional bool) (*WorkerAuthRepositoryStorage, error) {
```

Comment:
I think this will probably work -- I might need to modify the transaction interface so that the key id is passed into the function, I'm not sure how we'd get the worker otherwise since the value is opaque to the library. But that's pretty easy to do. Then when the function is called to create this you'd instead have the key ID passed in and from that you could look up the worker that it's associated with (if any).

Reply:
I had been assuming the calling function in Boundary (AuthorizeWorker) would do the lookup based on key ID to find/ create a worker before creating storage and passing in the worker.